src/PointwiseFunctions/Hydro/EquationsOfState/Tabulated3d.cpp

```diff
@@ -0,0 +1,435 @@
+    auto interpolated_state =
+        interpolator_.template interpolate<CsSquared>(weights);
+    get(cs2) = interpolated_state[0];
+
+  } else if constexpr (std::is_same_v<DataType, DataVector>) {
+    for (size_t s = 0; s < electron_fraction.size(); ++s) {
+      auto weights = interpolator_.get_weights(
+          get(log_temperature)[s], get(log_rest_mass_density)[s],
+          get(converted_electron_fraction)[s]);
+      auto interpolated_state =
+          interpolator_.template interpolate<CsSquared>(weights);
+      get(cs2)[s] = interpolated_state[0];
+    }
+  }
+
+  return cs2;
+}
+
+template <bool IsRelativistic>
+double Tabulated3D<IsRelativistic>::specific_internal_energy_lower_bound(
+    const double rest_mass_density, const double electron_fraction) const {
+  double converted_electron_fraction =
+      std::min(std::max(electron_fraction_lower_bound(), electron_fraction),
+               electron_fraction_upper_bound());
+
+  double log_rest_mass_density =
+      std::min(std::max(rest_mass_density_lower_bound(), rest_mass_density),
+               rest_mass_density_upper_bound());
+
+  log_rest_mass_density = log(log_rest_mass_density);
+
+  auto weights = interpolator_.get_weights(log(temperature_lower_bound()),
+                                           log_rest_mass_density,
+                                           converted_electron_fraction);
+  auto interpolated_state =
+      interpolator_.template interpolate<Epsilon>(weights);
+
+  return exp(interpolated_state[0]) + energy_shift_;
+}
+
+template <bool IsRelativistic>
+double Tabulated3D<IsRelativistic>::specific_internal_energy_upper_bound(
+    const double rest_mass_density, const double electron_fraction) const {
+  double converted_electron_fraction =
+      std::min(std::max(electron_fraction_lower_bound(), electron_fraction),
+               electron_fraction_upper_bound());
+
+  double log_rest_mass_density =
+      std::min(std::max(rest_mass_density_lower_bound(), rest_mass_density),
+               rest_mass_density_upper_bound());
+
+  log_rest_mass_density = log(log_rest_mass_density);
+
+  auto weights = interpolator_.get_weights(
+      log(upper_bound_tolerance_ * temperature_upper_bound()),
+      log_rest_mass_density, converted_electron_fraction);
+  auto interpolated_state =
+      interpolator_.template interpolate<Epsilon>(weights);
+
+  return exp(interpolated_state[0]) + energy_shift_;
+}
+
+template <bool IsRelativistic>
+Tabulated3D<IsRelativistic>::Tabulated3D(std::vector<double> electron_fraction,
+                                         std::vector<double> log_density,
+                                         std::vector<double> log_temperature,
+                                         std::vector<double> table_data,
+                                         double energy_shift,
+                                         double enthalpy_minimum) {
+  initialize(electron_fraction, log_density, log_temperature, table_data,
```

Comment:
`std::move` the `electron_fraction`, density, tem, table data here too. Squash right away :)

Reply:
Ha! Completely overlooked that one. :) Should be fixed now.